lnwallet/musig_session.go

```diff
@@ -0,0 +1,552 @@
+package lnwallet
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/shachain"
+)
+
+// commitType is an enum that denotes if this is the local or remote
+// commitment.
+type commitType uint8
+
+const (
+	// localCommit denotes that this a session for the local commitment.
+	localCommit commitType = iota
+
+	// remoteCommit denotes that this is a session for the remote
+	// commitment.
+	remoteCommit
+)
+
+// MusigPartialSig is a wrapper around the base musig2.PartialSignature type
+// that also includes information about the set of nonces used, and also the
+// signer. This allows us to implement the input.Signature interface, as that
+// requires the ability to perform abstract verification based on a public key.
+type MusigPartialSig struct {
+	// sig is the actual musig2 partial signature.
+	sig *musig2.PartialSignature
+
+	// signerNonce is the nonce used by the signer to generate the partial
+	// signature.
+	signerNonce lnwire.Musig2Nonce
+
+	// combinedNonce is the combined nonce of all signers.
+	combinedNonce lnwire.Musig2Nonce
+
+	// signerKeys is the set of public keys of all signers.
+	signerKeys []*btcec.PublicKey
+}
+
+// NewMusigPartialSig creates a new musig partial signature.
+func NewMusigPartialSig(sig *musig2.PartialSignature,
+	signerNonce, combinedNonce lnwire.Musig2Nonce,
+	signerKeys []*btcec.PublicKey) *MusigPartialSig {
+
+	return &MusigPartialSig{
+		sig:           sig,
+		signerNonce:   signerNonce,
+		combinedNonce: combinedNonce,
+		signerKeys:    signerKeys,
+	}
+}
+
+// FromWireSig maps a wire partial sig to this internal type that we'll use to
+// perform signature validation.
+func (p *MusigPartialSig) FromWireSig(sig *lnwire.PartialSigWithNonce,
+) *MusigPartialSig {
+
+	p.sig = &musig2.PartialSignature{
+		S: &sig.Sig,
+	}
+	p.signerNonce = sig.Nonce
+
+	return p
+}
+
+// ToWireSig maps the partial signature to something that we can use to write
+// out for the wire protocol.
+func (p *MusigPartialSig) ToWireSig() *lnwire.PartialSigWithNonce {
+	return &lnwire.PartialSigWithNonce{
+		PartialSig: lnwire.NewPartialSig(*p.sig.S),
+		Nonce:      p.signerNonce,
+	}
+}
+
+// Serialize serializes the musig2 partial signature. The serializing includes
+// the signer's public nonce _and_ the partial signature. The final signature
+// is always 98 bytes in length.
+func (p *MusigPartialSig) Serialize() []byte {
+	var b bytes.Buffer
+
+	_ = p.ToWireSig().Encode(&b) //nolint:errcheck
```

Comment:
If we propagate the error here, we'll need to change the `input.Signature` interface. As mentioned earlier, this can't return an error as it's currently defined.